synapse/events/utils.py

```diff
@@ -253,14 +245,76 @@ def _copy_field(src: JsonDict, dst: JsonDict, field: List[str]) -> None:
     sub_out_dict[key_to_move] = sub_dict[key_to_move]
 
 
+def _split_field(field: str) -> List[str]:
+    """
+    Split strings unescaped dots and removes escaping.
+
+    Args:
+        field: A string representing a path to a field.
+
+    Returns:
+        A list of nested fields to traverse.
+    """
+
+    # Convert the field and remove escaping:
+    #
+    # 1. "content.body.thing\.with\.dots"
+    # 2. ["content", "body", "thing\.with\.dots"]
+    # 3. ["content", "body", "thing.with.dots"]
+
+    result = []
+
+    # The current field and whether the previous character was the escape
+    # character (a backslash).
+    part = ""
+    escaped = False
+
+    # Iterate over each character, and decide whether to append to the current
+    # part (following the escape rules) or to start a new part (based on the
+    # field separator).
+    for c in field:
+        # If the previous character was the escape character (a backslash)
+        # then decide what to append to the current part.
+        if escaped:
+            if c in ("\\", "."):
+                # An escaped backslash or dot just gets added.
+                part += c
+            else:
+                # A character that shouldn't be escaped gets the backslash prepended.
+                part += "\\" + c
+            # This always resets being escaped.
+            escaped = False
+
+        # Otherwise, the previous character was not the escape character.
+        else:
+            if c == ".":
+                # The field separator creates a new part.
+                result.append(part)
+                part = ""
+            elif c == "\\":
+                # A backslash adds no characters, but starts an escape sequence.
+                escaped = True
+            else:
+                # Otherwise, just add the current character.
+                part += c
+
+    # Ensure the final part is included. If there's an open escape sequence
+    # it should be included.
+    if escaped:
+        part += "\\"
+    result.append(part)
+
+    return result
```

Comment:
I'm a little sad about the performance here, because the repeated string concatenation takes n\*\*2 time, unless Python is very clever.

If we wanted to, we could track the start and end positions of each part and substring them out. Then, if we saw any escapes, we could do some heavier processing.

However, field names are usually small so the current approach is probably okay.

Reply:
Building up a list containing chars to be concatenated all at once would be more optimal (in theory) than concatenating one char at a time. It might be best to benchmark the approaches to make sure I'm not talking nonsense though.

Reply:
(On string concatenation: you could try https://docs.python.org/3/library/io.html?highlight=stringio#io.StringIO which I think is more clever than just `+=`ing to a string.)

Reply:
I rewrote it to be quite different, let me know what you think.

Reply:
That's a nice way of doing the splitting. Looks right as far as I can tell.

Reply:
It took me a while to figure out a regular expression method that actually did what I needed...

I suppose I should compile the regexp before merging.